File: Frontend/lib/api/mock-interview.ts
import { apiGet, apiPost } from "./client"

export interface MockInterviewOption {
  id: string
  text: string
}

export interface MockInterviewQuestion {
  id: number
  question_text: string
  options: MockInterviewOption[]
  selected_option_id?: string | null
  topic: string
}

export interface MockInterviewSession {
  id: number
  career_context: string
  difficulty: string
  status: string
  score?: number | null
  total_questions: number
  started_at: string
  completed_at?: string | null
  questions: MockInterviewQuestion[]
}

export interface MockInterviewQuestionReview {
  id: number
  question_text: string
  options: MockInterviewOption[]
  selected_option_id?: string | null
  correct_option_id: string
  explanation: string
  topic: string
  source_resource_ids: string[]
  resource_url?: string | null
  resource_title?: string | null
}

export interface MockInterviewTopicPerformance {
  topic: string
  correct: number
  total: number
}

export interface MockInterviewResults {
  id: number
  career_context: string
  difficulty: string
  score: number
  total_correct: number
  total_questions: number
  performance_label: string
  topic_performance: MockInterviewTopicPerformance[]
  questions_review: MockInterviewQuestionReview[]
  next_best_action_text: string
  next_best_action_resource_id?: string | null
  next_best_action_url?: string | null
}

export async function setupMockInterview(career_context: string, difficulty: string): Promise<MockInterviewSession> {
  return apiPost<MockInterviewSession>("/mock-interviews/setup", {
    career_context,
    difficulty,
  })
}

export async function getMockInterview(sessionId: number | string): Promise<MockInterviewSession> {
  return apiGet<MockInterviewSession>(`/mock-interviews/${sessionId}`)
}

export async function submitMockInterviewAnswer(
  sessionId: number | string,
  questionId: number,
  selectedOptionId: string
): Promise<{ status: string }> {
  return apiPost<{ status: string }>(`/mock-interviews/${sessionId}/submit`, {
    question_id: questionId,
    selected_option_id: selectedOptionId,
  })
}

export async function completeMockInterview(sessionId: number | string): Promise<MockInterviewResults> {
  return apiPost<MockInterviewResults>(`/mock-interviews/${sessionId}/complete`, {})
}

export async function getMockInterviewResults(sessionId: number | string): Promise<MockInterviewResults> {
  return apiGet<MockInterviewResults>(`/mock-interviews/${sessionId}/results`)
}

File: BackEnd/schemas/mock_interview.py
"""Mock Interview Pydantic Schemas."""

from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime

class MockInterviewOption(BaseModel):
    id: str
    text: str

class MockInterviewQuestionSchema(BaseModel):
    id: str
    question: str
    options: List[MockInterviewOption]
    correct_option: str
    explanation: str
    topic: str
    source_resource_ids: List[str] = []

class MockInterviewGenerationResult(BaseModel):
    title: str
    career: str
    difficulty: str
    questions: List[MockInterviewQuestionSchema]

class MockInterviewSetupRequest(BaseModel):
    career_context: str
    difficulty: str

class MockInterviewQuestionResponse(BaseModel):
    id: int
    question_text: str
    options: List[MockInterviewOption]
    # No correct_option in the response to prevent cheating
    selected_option_id: Optional[str] = None
    topic: str

class MockInterviewSessionResponse(BaseModel):
    id: int
    career_context: str
    difficulty: str
    status: str
    score: Optional[float] = None
    total_questions: int
    started_at: datetime
    completed_at: Optional[datetime] = None
    questions: List[MockInterviewQuestionResponse]

class MockInterviewSubmitAnswerRequest(BaseModel):
    question_id: int
    selected_option_id: str

class MockInterviewResultTopic(BaseModel):
    topic: str
    correct: int
    total: int

class MockInterviewQuestionReview(BaseModel):
    id: int
    question_text: str
    options: List[MockInterviewOption]
    selected_option_id: Optional[str]
    correct_option_id: str
    explanation: str
    topic: str
    source_resource_ids: List[str] = []
    resource_url: Optional[str] = None
    resource_title: Optional[str] = None

class MockInterviewResultsResponse(BaseModel):
    id: int
    career_context: str
    difficulty: str
    score: float
    total_correct: int
    total_questions: int
    performance_label: str
    topic_performance: List[MockInterviewResultTopic]
    questions_review: List[MockInterviewQuestionReview]
    next_best_action_text: str
    next_best_action_resource_id: Optional[str] = None
    next_best_action_url: Optional[str] = None
